peer_review_data/main.py

```diff
@@ -99,103 +108,146 @@ def saveAssessmentsAndComments(
     for canvasAssessment in [CanvasAssessment(a) for a in
                              canvasAssessments]:
         if not canvasAssessment.isPeerReview:
-            LOGGER.info('Skipping non-peer-review assessment '
-                        f'({canvasAssessment.id}).')
-            continue
+            LOGGER.warning(f'Assessment ({canvasAssessment.id}) '
+                           'is NOT a peer-review.')
 
-        assessment: models.Assessment = None
+        assessment: Optional[models.Assessment] = None
+        problemType: str = None
+        problemObjectId: int = None
+        problemException: str = None
         try:
             assessment = \
                 models.Assessment.fromCanvasAssessment(canvasAssessment)
+
+            if assessment is None:
+                continue
+
             LOGGER.debug(f'Saving {assessment}…')
             assessment.save()
-        except Exception as e:
-            # XXX: Catches assessments referring to non-existent submissions!
-            LOGGER.warning(f'Error saving Assessment '
-                           f'({canvasAssessment.id}): {e}')
-            LOGGER.debug('Assessment data: ' + json.dumps(
-                dictSkipKeys(canvasAssessment, ['_requester']),
-                indent=2, default=str))
+        except Submission.DoesNotExist as e:
+            problemType = 'Submission'
+            problemObjectId = canvasAssessment.submissionId
+            problemException = e
+        except User.DoesNotExist as e:
+            problemType = 'Assessor'
+            problemObjectId = canvasAssessment.assessorId
+            problemException = e
```

Comment:
So as I noted before, these cases are happening pretty frequently. Do you have any ideas why they are happening? User leaves the course for instance? Just curious if you dug into this.

Reply:
No, I've not figured out why this happens.  I'd like to know why, but I'm somewhat comfortable with the way the code is handling this.  I've compared the results of this app with those from the Longhorn Open peer grading tool.  In an earlier iteration of my code, the results in the DB were several fewer peer reviews than the number Longhorn found.  I changed some things (don't skip reviews because they didn't weren't the type I thought they should be, etc.) and now this code finds only one less review than Longhorn does.

In a meeting with Instructure people, I mentioned that there were many instances of these missing entities.  They seemed surprised.  I asked whether it could be from the user leaving the course, but they were unsure.  They've asked me to send them example data, so I will do that soon.